static/js/reactions.js

```diff
@@ -0,0 +1,194 @@
+var reactions = (function () {
+var exports = {};
+
+var reactions_store = new Dict();
+
+function send_reaction_ajax(message_id, emoji_name, operation) {
+    var args = {
+        url: '/json/messages/' + message_id + '/emoji_reactions/' + emoji_name,
+        data: {},
+        success: function () {},
+        error: function (xhr) {
+          var response = channel.xhr_error_message("Error sending reaction", xhr);
+          blueslip.error(response);
+        }
+    };
+    if (operation === 'add') {
+        channel.put(args);
+    } else if (operation === 'remove') {
+        channel.del(args);
+    }
+}
+
+function message_reaction_on_click() {
+    // When a message's reaction is clicked,
+    // if the user has reacted to this message with this emoji
+    // the reaction is removed
+    // otherwise, the reaction is added
+    var emoji_name = $(this).attr('data-emoji-name');
+    var message_id = $(this).parent()[0].title;
+    var user_list = reactions_store.get(message_id).get(emoji_name);
+    var operation = 'remove';
+    if (user_list.indexOf(page_params.user_id) === -1) {
+        // User hasn't reacted with this emoji to this message
+        operation = 'add';
+    }
+    send_reaction_ajax(message_id, emoji_name, operation);
+}
+
+function check_reaction_operation(message_id, emoji_name) {
+    var operation = 'remove';
+    if (!reactions_store.has(message_id)) {
+      operation = 'add';
+    } else if (!reactions_store.get(message_id).has(emoji_name)) {
+      operation = 'add';
+    } else {
+        var user_list = reactions_store.get(message_id).get(emoji_name);
+        if (user_list.indexOf(page_params.user_id) === -1) {
+            // User hasn't reacted with this emoji to this message
+            operation = 'add';
+        }
+    }
+    return operation;
```

Comment:
This could probably be clearer if it eliminated the `operation` variable and just used `return` statements.